### PR TITLE
First pull request for problem_set_3 from Etienne

### PR DESCRIPTION
**Pull request by etiennedanis**
Hi,
The file to look at is: problem_set_3_Etienne.Rmd

I still have to work on the question 5 about the graph.
I initially wanted to finish all the questions of the problem_set_3 before pushing my commit but since I may have done a few mistakes in the first 4 questions, it may actually be great to get your feedback before I try to tackle the question 5.
Thank you very much in advance.
Best,
Etienne

### CODE REVIEW

**Comment by etiennedanis** [comment]
Hi, 
I found a very heavy way to get the set of 3 genes for each pair (plant + fungus) with the replicate values for the graphs. Could you please tell me how I could write a more elegant way?

I also add the issue of not being able to knit my rmd file into a html file. 
I managed to knit it as a pdf file but it is not ideal.
The error message I got is attached to this message.
If you know how I could fix it (I use Linux Mint), please let me know.

Thank you very much in advance!

Best,
error_message_in_rstudio_when_knitting_into_html

Etienne

**Comment by kriemo** [comment]
This looks like an issue with the fonts that are installed on your linux OS. 

Try the solution from this thread:
http://askubuntu.com/questions/449578/x11-font-adobe-helvetica-s-s-d-face-2-at-size-11-could-no

Also can you reply with the output from running:
sessionInfo() 
capabilities() 


capabilities('cairo')

**Comment by etiennedanis** [comment]
Yes. It seems to be the fonts.
I installed more fonts, I followed advices from different websites and nothing worked…

I saw the thread you sent me and preferred not to try their solution since it seemed to be designed for a 32-bit configuration and I have a 64-bit:
sudo apt-get install t1-xfree86-nonfree ttf-xfree86-nonfree ttf-xfree86-nonfree-syriac xfonts-75dpi xfonts-100dpi
But at this stage, since nothing else works, I will definitely try it and let you know.
Thank you very much for sending me this thread! I had forgotten about it.
If it does not work, I will try to reinstall R and Rstudio. I had a bunch of problems during the installation… (Rstudio could not find R…)

Here are the info you requested:

> sessionInfo()

R version 3.3.2 (2016-10-31)

Platform: x86_64-pc-linux-gnu (64-bit)

Running under: Linux Mint 18



locale:

[1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8

 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C

 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C



attached base packages:

[1] stats     graphics  grDevices utils     datasets  methods   base



loaded via a namespace (and not attached):

[1] tools_3.3.2



> capabilities()

       jpeg         png        tiff       tcltk         X11        aqua    http/ftp     sockets      libxml        fifo

       TRUE        TRUE        TRUE        TRUE        TRUE       FALSE        TRUE        TRUE        TRUE        TRUE

     cledit       iconv         NLS     profmem       cairo         ICU long.double     libcurl

       TRUE        TRUE        TRUE        TRUE        TRUE       FALSE        TRUE TRUE


> capabilities('cairo')

cairo

 TRUE

Let me know if you see anything else I could try.
Thank you so much in advance!

Best,

Etienne

**Comment by kriemo** [comment]
hmm ok. What is the output from X11.options()?

**Comment by etiennedanis** [comment]
Here is below the output for X11.options() (the $fonts looks pretty weird):

> X11.options()
$display
[1] ""

$width
[1] 8

$height
[1] 8

$pointsize
[1] 10

$bg
[1] "transparent"

$canvas
[1] "white"

$gamma
[1] 1

$colortype
[1] "true"

$maxcubesize
[1] 256

$fonts
[1] "-adobe-helvetica-%s-%s-*-*-%d-*-*-*-*-*-*-*" "-adobe-symbol-medium-r-*-*-%d-*-*-*-*-*-*-*"

$family
[1] "sans"

$xpos
[1] 0

$ypos
[1] NA

$title
[1] ""

$type
[1] "cairo"

$antialias
[1] "default"

File: problem_set_3_version_4.Rmd [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+raw_data | awk '$4 == "6hpi"' | sort -k6nr | head -n 1 | cut -f2
+```
+
+```{r}
+# translate to dplyr code
+
+filter(cleaned_data, time.point=="6hpi") %>% 
+  arrange(desc(value)) %>% 
+  select(plant) %>%  
+  filter(row_number()==1)
```
Comment (kriemo) [inline problem_set_3_version_4.Rmd]: In the future use the `slice()` dplyr function, whcih works much better than `row_number()` which is really a ranking function.

File: problem_set_3_version_4.Rmd [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+raw_data | awk '$4 == "18hpi"' | sort -k6nr | head -n 1 | cut -f2,3
+```
+
+```{r}
+# translate to dplyr code
+
+filter(cleaned_data, time.point=="18hpi") %>%
+  arrange(desc(value)) %>% 
+  select(plant, fungus) %>% 
+  filter(row_number()==1)
```
Comment (kriemo) [inline problem_set_3_version_4.Rmd]: ditto, use `slice()` instead in the future

File: problem_set_3_version_4.Rmd [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+#arrange(expr.diff.var) %>% 
+#slice(1:3) %>%
+#select(plant, fungus, gene.name, expr.diff.mean, expr.diff.var) %>% 
+#group_by(plant,fungus,gene.name) -> top3_genes
+#top3_genes %>% tbl_df() %>% print (n=12)
+```
+
+```{r}
+# Sort by expr.diff.mean last
+expr_data_mean %>% arrange(expr.diff.var) %>%
```
Comment (kriemo) [inline problem_set_3_version_4.Rmd]: Here you have arranged by `expr.diff.mean` but you are not arranged per group (i.e. for each gene and plant combination)

Compare your code to the following approach:
```r
expr_data_mean %>% arrange(expr.diff.var) %>% 
 arrange(desc(expr.diff.mean)) %>%
 group_by(plant, fungus) %>%
 slice(1:3) 
```

File: problem_set_3_version_4.Rmd [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+## Problem 4
+
+Now examine the above final data frame above and write a few sentences 
+putting inline code in least 3 places. 
+For example, There are `r nrow(mtcars)` rows of data in `mtcars`.
+
+
+```{r, echo = FALSE}
+
+filter(top3_genes, grepl("B12", plant)) %>%  filter(., grepl("A6", fungus)) %>%
```
Comment (kriemo) [inline problem_set_3_version_4.Rmd]: Use the grouping strategy above you will not need to filter each combination of plant and fungus as separate commands. Instead you could use a simple filter call as follows.

```r
filter(cleaned_data, gene.name %in% top3_genes$gene.name) 
```

File: problem_set_3_version_4.Rmd [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+  select(expr.diff.mean) %>%
+  slice(1) %>% format(round(1), nsmall = 1) -> top_gene_B12_K1
+
+filter(top3_genes, grepl("pps", plant)) %>%  filter(., grepl("K1", fungus)) %>% 
+  ungroup() %>% 
+  select(expr.diff.mean) %>%
+  slice(1) %>% format(round(1), nsmall = 1) -> top_gene_pps_K1
+
+#top3_genes_for_B12_A6
+top3_genes %>% subset(., plant =="B12" & fungus =="A6", select=c(gene.name)) %>%
```
Comment (kriemo) [inline problem_set_3_version_4.Rmd]: I would avoid using `subset()` and instead use dplyr's `filter()`, it's far easier to work with, and I suspect much quicker.

File: problem_set_3_version_4.Rmd [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+```{r preliminary_for_the_graphs}
+# 3 top genes per pair plant/fungus, 4 pairs, triplicates, 
+# 4 time points = 3 x 4 x 3 x 4 = 144
+# First, create for each pair plant/fungus, the list of top 3 genes 
+#differentially expressed and save the info in different files
+cleaned_data %>% mutate(time.point = str_replace(time.point, 'hpi', '')) %>%
+  select(gene.name, plant, fungus, rep, time.point, value) %>%
+  group_by(plant,fungus,gene.name)-> data_selected
+
+top3_genes %>% subset(., plant =="B12" & fungus =="A6",
```
Comment (kriemo) [inline problem_set_3_version_4.Rmd]: see comments above for advice on how to perform these operations using only a few lines of code.

File: problem_set_3_version_4.Rmd [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+filter(data_selected, gene.name %in% top3_genes_for_pps_K1$gene.name) -> 
+  data_filtered_by_gene_name_pps_K1
+filter(data_filtered_by_gene_name_pps_K1, plant %in% top3_genes_for_pps_K1$plant) -> 
+  data_filtered_by_gene_name_and_plant_pps_K1
+filter(data_filtered_by_gene_name_and_plant_pps_K1, 
+       fungus %in% top3_genes_for_pps_K1$fungus) ->
+  data_for_pps_and_K1_pair
+
+#Combine all the dataframes in 1 for the graphs
+all_data_for_graphs <-
```
Comment (kriemo) [inline problem_set_3_version_4.Rmd]: Good use of `rbind()`, although again unnecessary if you use proper grouping and filtering. FYI, dplyr has a very useful `bind_rows()` and `bind_cols()` which provide similar functionality to the base R functions.

File: problem_set_3_version_4.Rmd [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+  ggtitle('Gene expression profile of the top 3 differentially 
+          expressed genes per pair plant/fungus over the time course')
+```
+
+7. Use the "black & white" theme
+
+```{r question_5_7}
+# Theme black and white using theme_bw() (and also theme cowplot, theme_cowplot())
+
+library(cowplot)
```
Comment (kriemo) [inline problem_set_3_version_4.Rmd]: Nice work.

File: problem_set_3_version_4.Rmd [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+  scale_shape_manual(values=c(21,24,8)) + 
+  geom_smooth(method = 'lm', se = F) +
+  ggtitle('Gene expression profile of the top 3 differentially 
+          expressed genes per pair plant/fungus over the time course') + 
+  theme_bw() + theme_cowplot()
+```
+
+Extra credit: add error bars to the plot (use `geom_errorbar`).
+
+```{r 5.8 extra credit - variant 1}
```
Comment (kriemo) [inline problem_set_3_version_4.Rmd]: Excellent work.

**Comment by etiennedanis** [comment]
Hi Kent,

I think that the recommendation that you sent me previously was right 
since now it is working.

I typed in bash:

sudo apt-get install t1-xfree86-nonfree ttf-xfree86-nonfree 
ttf-xfree86-nonfree-syriac xfonts-75dpi xfonts-100dpi

Unfortunately, initially, it did not help with my problem of knitting in 
html.

So I tried the other solution I found on line, which stated to change 
the setting for X11 'type' from xlib to cairo as it is recommended on 
several sites by adding the following lines in .Rprofile in the working 
directory or $HOME directory:

# default X11() setting
setHook(packageEvent("grDevices", "onLoad"),
         function(...) grDevices::X11.options(width = 8, height = 8, 
xpos = 0, pointsize = 10, type='cairo'))
options(device='x11')

Since I did not have any .Rprofile, I created one and edited it with the 
lines above.
I restarted Rstudio and it did change the type of X11 (checked using 
X11.options()) but I still could not knit in html.

Thinking that I may be missing some packages or libraries, I went to the 
Synaptic Package Manager in Linux Mint and installed all the packages 
and libraries containing 'X11' in their title and which based on their 
description seemed to be potentially involved in the visualization of 
graphs with X11.
After that, I restarted Rstudio and the knitting in html still did not work.

I went back to Google and found this thread:
https://forum.qiime2.org/t/error-in-sequence-quality-control-in-moving-pictures-tutorial/225/5 


Suggesting that after running:

sudo apt-get install t1-xfree86-nonfree ttf-xfree86-nonfree 
ttf-xfree86-nonfree-syriac xfonts-75dpi xfonts-100dpi

the computer should be restarted.
So I restarted my computer and finally it worked!
It seems that this command line followed by restarting the computer may 
be the answer. Though I did a lot of other changes and installed other 
packages and libraries that could have been taken into account only 
after restarting the computer so I still have a doubt for the real 
solution here.
But the good news is that now (so far) it works.

Thank you again for all the great suggestions!
Without them, I would have been spent much more time on it.

Best,

Etienne

**Comment by etiennedanis** [comment]
Hi Kent,

I just submitted my last version of problem_set_3 with all your suggestions.

Thanks a lot for all your input!
I still have a question concerning the problem_set_3.

To get the top 3 genes for each pair 'plant + fungus', you recommended the following way:

expr_data_mean %>% arrange(expr.diff.var) %>% 
 arrange(desc(expr.diff.mean)) %>%
 group_by(plant, fungus) %>%
 slice(1:3) -> top3_genes

At that stage, the code works well. It gives me the top 3 genes for each pair.
Then to extract the replicate data for each top 3 genes for each pair, you recommended to filter the data the following way:

filter(cleaned_data, gene.name %in% top3_genes$gene.name)  

It works but weirdly, when I do that, instead of giving me the top 3 genes for each pair, it gives me the top 4 genes.
If I change the line: slice(1:3) by: slice(1:2) then I get the top 3 genes later on after the filter line.
Is it normal? Am I forgetting some intermediate processing? Is it a bug in my version of Rstudio...?

Thank you very much in advance.

Etienne

**Comment by kriemo** [comment]
The reason that you are seeing 4 genes after filtering is because there are 4 genes found in the top 3 table that you've generated. If you looks at the `top_genes0` tbl in your code, you'll see that there are 4 unique genes. When you plot the data, all 4 genes of these genes are plotted for all conditions. This is desirable as it allows you to see the gene expression differences in all conditions. 

```r
unique(top3_genes0$gene.name)
[1] "bgh01942" "bgh00377" "bgh05405" "bgh02634"
```

@etiennedanis

**Comment by kriemo** [comment]
Strong work 👍

**Pull request closed without merging**